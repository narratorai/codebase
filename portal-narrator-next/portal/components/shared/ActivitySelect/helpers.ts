import { isEmpty } from 'lodash'

export const makeRowsOfThree = (items: any) => {
  const rows = []

  for (let i = 0; i < items?.length || 0; i += 3) {
    const row = [items[i]]
    if (!isEmpty(items[i + 1])) {
      row.push(items[i + 1])
    }
    if (!isEmpty(items[i + 2])) {
      row.push(items[i + 2])
    }

    rows.push(row)
  }

  return rows
}
