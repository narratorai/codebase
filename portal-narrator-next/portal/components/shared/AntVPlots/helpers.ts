import { DualAxesConfig } from '@ant-design/charts'
import ValueFormatter from 'components/shared/DataTable/ValueFormatter'
import { ICompany } from 'graph/generated'
import { produce } from 'immer'
import { each, includes, isEmpty, isFinite, isString, keys, map } from 'lodash'
import { TooltipItem } from 'node_modules/@antv/g2/lib/interface'

import { MULTI_Y_AXIS_PLOTS } from './constants'
import { AntVPlotConfigs, AntVPlotTypes } from './interfaces'

// Do not format large numbers in tooltip
const FORMAT_OPTIONS = {
  formatLargeNumbers: true,
  truncateLimit: 30, // hard coded for now - could make dynamic though
}

function getXAxisConfig(valueFormatter: ValueFormatter, timezone: string, config?: AntVPlotConfigs) {
  const xAxisFormatter = (value: string | number) => {
    const format = config?.xAxis?.label?.narrator_format

    if (format) return valueFormatter.formatValue(format, value, timezone, FORMAT_OPTIONS)
    return value
  }

  return {
    ...config?.xAxis,
    label: {
      ...config?.xAxis?.label,
      formatter: xAxisFormatter,
    },
  }
}

function getYAxisConfig(
  valueFormatter: ValueFormatter,
  timezone: string,
  hasMultiYAxes: boolean,
  config?: AntVPlotConfigs
) {
  const yAxisFormatter = (value: string | number) => {
    const format = config?.yAxis?.label?.narrator_format

    if (format) return valueFormatter.formatValue(format, value, timezone, FORMAT_OPTIONS)
    return value
  }

  const yMultiAxesFormatter = (value: string | number, format: string) => {
    if (format) return valueFormatter.formatValue(format, value, timezone, FORMAT_OPTIONS)
    return value
  }

  if (hasMultiYAxes) {
    const yAxisKeys = keys(config?.yAxis)
    const axisConfig: DualAxesConfig['yAxis'] = {}

    each(yAxisKeys, (yAxisKey) => {
      const axisConfig = config?.yAxis[yAxisKey]
      const format = axisConfig?.label?.narrator_format

      axisConfig[yAxisKey] = {
        ...axisConfig,
        label: {
          ...axisConfig?.label,
          formatter: (value: string | number) => yMultiAxesFormatter(value, format),
        },
      }
    })

    return axisConfig
  }

  return {
    ...config?.yAxis,
    label: {
      ...config?.yAxis?.label,
      formatter: yAxisFormatter,
    },
  }
}

function getLabelFormatter(valueFormatter: ValueFormatter, timezone: string, config?: AntVPlotConfigs) {
  const labelFormat = config?.label?.narrator_format
  const labelFormatField = config?.label?.format_field

  const labelFormatter = (data: Record<string, string | number>) => {
    if (labelFormatField && labelFormat) {
      const shownValue = data?.[labelFormatField]

      if (shownValue) {
        const formattedValue = valueFormatter.formatValue(labelFormat, shownValue, timezone, FORMAT_OPTIONS)
        // make sure the formattedValue is a number or string before returning
        if (isFinite(formattedValue) || isString(formattedValue)) {
          return formattedValue
        }
      }
    }

    // default show nothing (label format was passed, but was not correctly formatted)
    return null
  }

  if (!labelFormat || !labelFormatField) return config?.label

  return {
    ...config.label,
    formatter: labelFormatter,
  }
}

function getSliderConfig(valueFormatter: ValueFormatter, timezone: string, config?: AntVPlotConfigs) {
  const sliderFormat = config?.slider?.narrator_format
  const sliderFormatter = (value: string | number) => {
    if (sliderFormat) {
      return valueFormatter.formatValue(sliderFormat, value, timezone, FORMAT_OPTIONS)
    }

    return value
  }

  if (!config?.slider) return undefined

  return {
    ...(config.slider as object),
    formatter: sliderFormatter,
  }
}

function getTooltipConfig(
  valueFormatter: ValueFormatter,
  timezone: string,
  config?: AntVPlotConfigs,
  getTooltipData?: (data: any) => void
) {
  // for single axis
  const tooltipFormat = config?.tooltip?.narrator_format

  // for multi axes
  const tooltipsFormats = config?.tooltip?.narrator_formats

  const getTooltipItems = (items: TooltipItem[]) => {
    getTooltipData?.(items[0]?.data)

    // single axis
    if (tooltipFormat) {
      return map(items, (item) => {
        const formattedValue = valueFormatter.formatValue(tooltipFormat, item.value, timezone)

        return {
          ...item,
          data: {
            ...item.data,
            // only tiny-area seems to need this, but it will not show the
            // updated value unless the data is formatted as well
            [item.name]: formattedValue,
          },
          value: formattedValue,
        }
      })
    }

    // multi axes
    if (tooltipsFormats) {
      return map(items, (item, index) => ({
        ...item,
        value: valueFormatter.formatValue(tooltipsFormats[index], item.value, timezone),
      }))
    }

    return items
  }

  return {
    ...config?.tooltip,
    customItems: getTooltipItems,
  }
}

function getMeta(valueFormatter: ValueFormatter, timezone: string, config?: AntVPlotConfigs) {
  if (isEmpty(config?.meta)) {
    return {}
  }

  return produce(config.meta, (metaDraft) => {
    Object.keys(metaDraft).forEach((key) => {
      const format = metaDraft[key].narrator_format

      if (format) {
        metaDraft[key].formatter = (value: string | number) =>
          valueFormatter.formatValue(format, value, timezone, FORMAT_OPTIONS)
      }
    })
  })
}

function getGeometryOptionsConfig(
  valueFormatter: ValueFormatter,
  timezone: string,
  hasMultiYAxes: boolean,
  config?: AntVPlotConfigs
) {
  const formatter = (data: Record<string, string | number>, format: string, field: string) => {
    const value = data[field]

    const valueIsValid = isFinite(value) || !isEmpty(value)
    if (hasMultiYAxes && format && valueIsValid) {
      return valueFormatter.formatValue(format, value, timezone, FORMAT_OPTIONS)
    }

    return value
  }

  if (hasMultiYAxes && !isEmpty(config?.geometryOptions)) {
    return map(config?.geometryOptions, (option) => {
      const field = option.label?.format_field
      const format = option.label?.narrator_format

      if (field && format) {
        return {
          ...option,
          label: {
            ...option.label,
            formatter: (data: Record<string, string | number>) => formatter(data, format, field),
          },
        }
      }

      // if no format or format field return the original option
      return option
    })
  }

  return undefined
}

export function getPlotConfig(
  company: ICompany,
  timezone: string,
  height: number | undefined,
  config?: AntVPlotConfigs,
  type?: AntVPlotTypes,
  getTooltipData?: (data: any) => void
) {
  const valueFormatter = new ValueFormatter(undefined, company.currency_used || 'USD')
  const hasMultiYAxes = includes(MULTI_Y_AXIS_PLOTS, type)

  return {
    ...config,
    meta: getMeta(valueFormatter, timezone, config),
    xAxis: getXAxisConfig(valueFormatter, timezone, config),
    yAxis: getYAxisConfig(valueFormatter, timezone, hasMultiYAxes, config),
    label: getLabelFormatter(valueFormatter, timezone, config),
    slider: getSliderConfig(valueFormatter, timezone, config),
    tooltip: getTooltipConfig(valueFormatter, timezone, config, getTooltipData),
    geometryOptions: getGeometryOptionsConfig(valueFormatter, timezone, hasMultiYAxes, config),
    height,
  }
}
