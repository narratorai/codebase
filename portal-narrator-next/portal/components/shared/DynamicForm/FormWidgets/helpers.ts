import { each, map, includes, isEmpty, compact } from 'lodash'

export interface TreeItem {
  value: string
  title: string
  selectable?: boolean
  children?: TreeItem[]
}

interface CleanTreeItemProps {
  treeItem: TreeItem
  enumValues: string[]
}

export const cleanTreeItem = ({ treeItem, enumValues }: CleanTreeItemProps) => {
  const updatedChildren: TreeItem[] = []
  // if item has children
  if (treeItem.children) {
    // loop through all children
    each(treeItem.children, (subTreeItem) => {
      // if this child has children
      // clean each of those tree items (all the way down)
      if (!isEmpty(subTreeItem.children)) {
        const nestedChildren = cleanTreeItem({ treeItem: subTreeItem, enumValues })

        // don't include empty dropdowns
        if (nestedChildren?.selectable === false && !isEmpty(nestedChildren?.children)) {
          updatedChildren.push(nestedChildren)
        }
      }

      // and only keep children whose values belong to enumValues
      if (includes(enumValues, subTreeItem.value)) {
        updatedChildren.push(subTreeItem)
      }
    })
  }

  const updatedTreeItem = {
    ...treeItem,
    children: updatedChildren,
  }

  if (updatedTreeItem.selectable === false && isEmpty(updatedChildren)) {
    return undefined
  }

  return updatedTreeItem
}

interface CleanTreeDataProps {
  treeData: TreeItem[]
  enumValues: string[]
}

export const cleanTreeData = ({ treeData, enumValues }: CleanTreeDataProps) => {
  return compact(map(treeData, (treeItem) => cleanTreeItem({ treeItem, enumValues })))
}
