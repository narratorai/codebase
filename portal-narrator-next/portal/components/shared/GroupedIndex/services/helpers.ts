import _ from 'lodash'
import moment from 'moment-timezone'
import {
  GROUPING_RECENT,
  GROUPING_RECENT_NEW_LABEL,
  GROUPING_RECENT_NOT_NEW_LABEL,
  GROUPING_NULL_LABEL,
  RECENT_CALCULATION_DAY_COUNT,
} from './constants'

export interface IGroupConfig {
  label: string
  group: string
  pathToValue: string
  groupLabelOverrides?: {
    key: string
    label: string
  }[]
  addToGroup?: {
    from: string
    to: string
  }[]
}

/////
// isItemNew()
//
// Calculate whether an item in the index is classified as "new"
// based on the number of days since it was created
/////
interface IsItemNewArgs<T> {
  item: T
  pathToTimestamp?: string
  tz?: string
}
export const isItemNew = <T>({
  item,
  pathToTimestamp = 'created_at',
  tz = 'America/New York',
}: IsItemNewArgs<T>): boolean => {
  const timestamp = _.get(item, pathToTimestamp)
  if (!timestamp) {
    return false
  }

  const daysSinceCreated = moment(new Date()).diff(moment.tz(timestamp, tz), 'days')
  return daysSinceCreated <= RECENT_CALCULATION_DAY_COUNT
}

/////
// getAllPossibleGroupKeys()
//
// antd's <Collapse> component expects `defaultActiveKey` prop to have all the open panel keys in it
// Since we're using a <Radio> to decide with Collapse headers to show, I need to include
// every possible panel key so they are all open by defualt when swapping between group types
// https://ant.design/components/collapse/
/////
interface GetAllPossibleGroupKeysArgs<T> {
  items: T[]
  groupConfigs: IGroupConfig[]
}
export const getAllPossibleGroupKeys = <T>({ items, groupConfigs }: GetAllPossibleGroupKeysArgs<T>): string[] => {
  const groupValues = _.reduce(
    groupConfigs,
    (result: string[], config) => {
      const uniqueValues = _.compact(_.uniq(_.map(items, config.pathToValue)))
      const overrideLabels = _.map(config.groupLabelOverrides, 'label')
      return [...result, ...uniqueValues, ...overrideLabels]
    },
    []
  )
  return [...groupValues, GROUPING_RECENT_NEW_LABEL, GROUPING_RECENT_NOT_NEW_LABEL, GROUPING_NULL_LABEL]
}

/////
// groupItems()
//
// Take list of items, and _.group them based on a selectedGroup
// Includes support for GROUPING_RECENT and a list of user defined groupConfigs
/////
interface GroupItemsArgs<T> extends Pick<IsItemNewArgs<T>, 'pathToTimestamp' | 'tz'> {
  items: T[]
  groupConfigs: IGroupConfig[]
  selectedGroup: string
}
interface GroupItemsResponse<T> {
  [key: string]: T[]
}

export const groupItems = <T>({
  items,
  groupConfigs,
  selectedGroup,
  pathToTimestamp,
  tz,
}: GroupItemsArgs<T>): GroupItemsResponse<T> | null => {
  if (selectedGroup === GROUPING_RECENT) {
    const groupedByNew = _.groupBy(items, (item) => isItemNew({ item, pathToTimestamp, tz }))
    return {
      [GROUPING_RECENT_NEW_LABEL]: groupedByNew['true'],
      [GROUPING_RECENT_NOT_NEW_LABEL]: groupedByNew['false'],
    }
  }

  const selectedGroupConfig = _.find(groupConfigs, ['group', selectedGroup])

  if (selectedGroupConfig) {
    // make sure undefined items are last in the list
    const sortedItems = _.sortBy(items, (item) => _.isEmpty(_.get(item, selectedGroupConfig.pathToValue)))

    const groupedValues = _.groupBy(sortedItems, selectedGroupConfig.pathToValue)
    // Allow overwriting the default _.groupBy keys with new labels
    // ex: validated: true/false --> validated: "Validated"/"Not Validated"
    if (selectedGroupConfig.groupLabelOverrides || selectedGroupConfig.addToGroup) {
      const groupedValuesWithOverride: any = {}
      _.forEach(groupedValues, (value, key) => {
        // first check for group label overrides (i.e. 'undefined' -> 'Not Materialized')
        const override = _.find(selectedGroupConfig.groupLabelOverrides, ['key', key])
        if (override) {
          groupedValuesWithOverride[override.label] = [...(groupedValuesWithOverride[override.label] || []), ...value]
        } else {
          groupedValuesWithOverride[key] = [...(groupedValuesWithOverride[key] || []), ...value]
        }

        // Now check for addToGroup (some items should be in more than one group i.e. 'gsheets' in 'gsheets' and 'materialized_view')
        // (this is why we splat value into previous values)
        const groupToAdd = _.find(selectedGroupConfig.addToGroup, ['from', key])
        if (groupToAdd) {
          groupedValuesWithOverride[groupToAdd.to] = [...(groupedValuesWithOverride[groupToAdd.to] || []), ...value]
        }
      })

      return groupedValuesWithOverride
    }

    // Overwrite "null" to be "Uncategorized"
    if (_.includes(_.keys(groupedValues), 'null')) {
      return {
        ..._.omit(groupedValues, 'null'),
        [GROUPING_NULL_LABEL]: groupedValues['null'],
      }
    }

    return groupedValues
  }

  return null
}

/////
// searchItems()
//
// Search through a list of items matching the serachValue to values
// from a list of _.get searchablePaths
//
// FIXME - consider moving this type of logic to graph?
/////
export type SearchablePathFunction = (item: any, searchValue: string) => boolean

interface SearchItemsArgs<T> {
  items: T[]
  searchablePaths: (string | SearchablePathFunction)[]
  searchValue: string
}

export const searchItems = <T>({ items, searchablePaths, searchValue }: SearchItemsArgs<T>): T[] => {
  if (!searchValue) {
    return items
  }

  return _.filter(items, (item: T) =>
    _.some(searchablePaths, (path) => {
      if (_.isString(path)) {
        const value = _.get(item, path)

        return _.includes(_.lowerCase(value), _.lowerCase(searchValue))
      }

      if (_.isFunction(path)) {
        return path(item, searchValue)
      }
    })
  )
}
