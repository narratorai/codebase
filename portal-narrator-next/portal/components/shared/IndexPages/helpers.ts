import _ from 'lodash'
import { ITag } from './interfaces'
import { NON_SHARED_TAGS } from './constants'
import {
  IListCompanyTagsQuery,
  ICreateCompanyTagMutationVariables,
  IMetric_Tags,
  IDataset_Tags,
  INarrative_Tags,
} from 'graph/generated'
import { reportError } from 'util/errors'

// We store each user's favorited and recently viewed items (Dataset, Kpi ...) in graph as tags.
// Each CompanyTag is specific to a company and to a user (every user in every company has its own 'favorite' tag)
// If it's the first time we associate a tag to an item (Dataset, Kpi ...) - instantiate the company tag
interface EnsureCompanyTagForUserProps {
  tagName: string // i.e. "favorite" or "recently_viewed"
  tags: ITag[]
  companyId: string
  userId: string
  createCompanyTag: ({ variables }: { variables: ICreateCompanyTagMutationVariables }) => void
}

export const ensureCompanyTagForUser = async ({
  tagName,
  tags,
  companyId,
  userId,
  createCompanyTag,
}: EnsureCompanyTagForUserProps) => {
  const foundTag = _.find(tags, (tag) => tag.tag == tagName)
  if (!foundTag) {
    try {
      const response = await createCompanyTag({
        variables: { tag: tagName, company_id: companyId, user_id: userId },
      })
      const tagId = _.get(response, 'data.insert_company_tags_one.id')
      return tagId
    } catch (e) {
      // nothing for user to be notified about so just report it
      return reportError('Failed to create a new company tag', e as Error)
    }
  }

  return foundTag.id
}

export const getSharedCompanyTags = (companyTags?: IListCompanyTagsQuery['company_tags']) =>
  _.filter(companyTags, (tag) => !_.includes(NON_SHARED_TAGS, tag.tag))

export const isSharedTag = (tag: IMetric_Tags | IDataset_Tags | INarrative_Tags) => {
  // make sure there is a tag to check against
  if (_.isEmpty(tag?.company_tag?.tag)) {
    return false
  }

  return !_.includes(NON_SHARED_TAGS, tag?.company_tag?.tag)
}

export const hasSharedTags = (tags: IMetric_Tags[] | IDataset_Tags[] | INarrative_Tags[] | undefined) =>
  !_.isEmpty(_.filter(tags || [], (tag) => isSharedTag(tag as IMetric_Tags | IDataset_Tags)))
