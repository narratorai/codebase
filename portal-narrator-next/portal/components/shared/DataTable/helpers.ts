import { each, keys, map, find, isEmpty, get, kebabCase } from 'lodash'
import copy from 'copy-to-clipboard'
import { downloadCsv } from 'util/download'
import { csvValueSanitizer } from 'util/helpers'

import { ColDef, ColGroupDef } from '@ag-grid-community/core'
import { RowData, ITableData } from 'components/shared/DataTable/interfaces'

// ag-grid will return {field: value} for each cell
// but we want to copy {headerName: value} so it matches
// what the user sees in the dataset table
export const formatRowData = ({ rowData, columnDefs }: { rowData: RowData; columnDefs?: ColDef | ColGroupDef }) => {
  // shouldn't happen, but if there are no column defs
  // return rowData as is
  if (!columnDefs) {
    return rowData
  }

  const formattedRowData: Record<string, string | number | null> = {}

  // loop through each rowData key (field)
  each(keys(rowData), (columnField) => {
    // use the field to find the columnDef
    const foundColumnDef = find(columnDefs, ['field', columnField])

    // use the columnDef to get the correct headerName
    if (foundColumnDef) {
      const headerName = foundColumnDef.headerName
      // and keep the same value initially passed in rowData
      const cellValue = rowData[columnField]
      formattedRowData[headerName] = cellValue
    }
  })

  return formattedRowData
}

// handle formatting row data when multiple rows are selected
export const formatMultipleRowData = ({
  selectedRowsData,
  columnDefs,
}: {
  selectedRowsData: RowData[]
  columnDefs?: ColDef | ColGroupDef
}) => {
  return map(selectedRowsData, (rowData) => formatRowData({ rowData, columnDefs }))
}

// note: rows should be pre-formatted
// note: this is technically tab separated values
// b/c it seems to work better when pasting into google sheets
export const formatCsvData = ({ rows }: { rows: RowData[] }) => {
  if (!rows || isEmpty(rows)) {
    return null
  }

  const headerKeys = keys(rows[0])
  // separate headers by tab
  const csvHeader = headerKeys.join('\t')
  // separate each row value by tab, then add a line break for next row
  const csvRows = rows.map((row) => headerKeys.map((key) => row[key]).join('\t')).join('\n')

  return csvHeader + '\n' + csvRows
}

interface CopyJsonCsvProps {
  selectedRows?: RowData[]
  rowData?: RowData
  columnDefs?: ColDef | ColGroupDef
}

export const copyJson = ({ selectedRows, rowData, columnDefs }: CopyJsonCsvProps) => {
  // check if only one row was selected
  if (isEmpty(selectedRows) && !isEmpty(rowData)) {
    return copy(JSON.stringify(formatRowData({ rowData, columnDefs })))
  }

  // otherwise multiple rows were selected
  if (!isEmpty(selectedRows) && selectedRows) {
    return copy(JSON.stringify(formatMultipleRowData({ selectedRowsData: selectedRows, columnDefs })))
  }

  // shouldn't happen, but safety return
  return null
}

export const copyCsv = ({ selectedRows, rowData, columnDefs }: CopyJsonCsvProps) => {
  // check if only one row was selected
  if (isEmpty(selectedRows) && !isEmpty(rowData)) {
    const formattedRowData = formatRowData({ rowData, columnDefs })
    const csvData = formatCsvData({ rows: [formattedRowData] })
    if (csvData) {
      // use navigator instead of copy for formatting purposes
      return navigator.clipboard.writeText(csvData)
    }
  }

  // otherwise multiple rows were selected
  if (!isEmpty(selectedRows) && selectedRows) {
    const formattedSelectedRowsData = formatMultipleRowData({ selectedRowsData: selectedRows, columnDefs })
    const csvData = formatCsvData({ rows: formattedSelectedRowsData })
    if (csvData) {
      // use navigator instead of copy for formatting purposes
      return navigator.clipboard.writeText(csvData)
    }
  }
}

interface DownloadTableDataAsCsvProps {
  title?: string
  tableData: ITableData
}

export const downloadTableDataAsCsv = ({ tableData, title }: DownloadTableDataAsCsvProps) => {
  const fileName = `${kebabCase(title) || 'table'}-${get(tableData, 'retrieved_at', 'table').slice(0, 19)}`

  // use column names as keys to maintain column orders
  // when mapping through row values
  const tableKeys: string[] = tableData.columns.map((column) => column.name)
  const csvHeader = tableKeys.map((tableKey) => csvValueSanitizer(tableKey))

  const csvData =
    csvHeader +
    '\n' +
    tableData.rows.map((row) => tableKeys.map((key) => csvValueSanitizer(row[key])).join(',')).join('\n')

  return downloadCsv({ csvData, fileName })
}
