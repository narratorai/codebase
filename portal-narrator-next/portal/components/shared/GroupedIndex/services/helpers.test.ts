import _ from 'lodash'
import { getAllPossibleGroupKeys, groupItems, searchItems } from './helpers'

const mockItems = [
  {
    name: 'Fancy Name',
    description: 'This is a description',
    category: 'fancy',
    validated: false,
    null_category: null,
    other: 'Should not be searched',
    nested: {
      value: 'nested is fun!',
    },
    only_in_fancy_name: 'better only see me in fancy!',
  },
  {
    name: 'Hipster Ipsum',
    description: 'Banjo anim ramps chia VHS batch flexitarian bespoke meh pinterest esse beard.',
    category: 'hipster',
    validated: false,
    null_category: null,
    other: 'Should not be searched',
    nested: {
      value: 'Green juice cillum ut, 3 wolf moon',
    },
  },
]

describe('GroupedIndex/services/helpers', () => {
  describe('#getAllPossibleGroupKeys', () => {
    it('gets all group keys including GROUP_RECENT keys', () => {
      expect(
        getAllPossibleGroupKeys({
          items: mockItems,
          groupConfigs: [{ label: 'category', group: 'category group', pathToValue: 'category' }],
        })
      ).toMatchSnapshot()
    })

    describe('with groupLabelOverrides', () => {
      it('gets all group keys from overrides as well', () => {
        expect(
          getAllPossibleGroupKeys({
            items: mockItems,
            groupConfigs: [
              {
                label: 'validated',
                group: 'validated',
                pathToValue: 'validated',
                groupLabelOverrides: [
                  { key: 'true', label: 'Validated' },
                  { key: 'false', label: 'Not Validated' },
                ],
              },
            ],
          })
        ).toMatchSnapshot()
      })
    })
  })

  describe('#groupItems', () => {
    it('groups items', () => {
      expect(
        groupItems({
          items: mockItems,
          groupConfigs: [{ label: 'category', group: 'category group', pathToValue: 'category' }],
          selectedGroup: 'category group',
        })
      ).toMatchSnapshot()
    })

    describe('with null group values', () => {
      it('includes items with null groups into an "Uncategorized" group', () => {
        expect(
          groupItems({
            items: mockItems,
            groupConfigs: [{ label: 'null_category', group: 'null_category', pathToValue: 'null_category' }],
            selectedGroup: 'null_category',
          })
        ).toMatchSnapshot()
      })

      it('should put falsy values at the bottom', () => {
        const groupedItems = groupItems({
          items: mockItems,
          groupConfigs: [
            { label: 'only_in_fancy_name', group: 'only_in_fancy_name', pathToValue: 'only_in_fancy_name' },
          ],
          selectedGroup: 'only_in_fancy_name',
        })

        expect(_.last(_.keys(groupedItems))).toBe('undefined')
      })
    })

    describe('with groupLabelOverrides', () => {
      it('only overrides the default group names from _.groupBy', () => {
        expect(
          groupItems({
            items: mockItems,
            groupConfigs: [
              {
                label: 'validated',
                group: 'validated',
                pathToValue: 'validated',
                groupLabelOverrides: [
                  { key: 'true', label: 'Validated' },
                  { key: 'false', label: 'Not Validated' },
                ],
              },
            ],
            selectedGroup: 'validated',
          })
        ).toMatchSnapshot()
      })

      it('adds values from one group to both when passed addToGroup', () => {
        expect(
          groupItems({
            items: mockItems,
            groupConfigs: [
              {
                label: 'category',
                group: 'category',
                pathToValue: 'category',
                addToGroup: [{ from: 'fancy', to: 'hipster' }],
              },
            ],
            selectedGroup: 'category',
          })
        ).toMatchSnapshot()
      })
    })
  })

  describe('#searchItems', () => {
    describe('with lower case values', () => {
      it('returns filtered items with mixed casing', () => {
        expect(
          searchItems({
            items: mockItems,
            searchablePaths: ['name', 'description'],
            searchValue: 'hipster',
          })
        ).toEqual([mockItems[1]])
      })
    })

    describe('with nested searchablePaths', () => {
      it('returns filtered items', () => {
        expect(
          searchItems({
            items: mockItems,
            searchablePaths: ['nested.value'],
            searchValue: 'nested is',
          })
        ).toEqual([mockItems[0]])
      })
    })

    describe('with searchablePath as a function', () => {
      it('returns an item when conditions are met', () => {
        const conditionalFunction = (item: any, searchValue: string) => {
          const category = _.get(item, 'category')
          const searchValueInLabel = _.includes(_.lowerCase(category), _.lowerCase(searchValue))

          if (searchValueInLabel && item.description === 'This is a description') {
            return true
          }

          return false
        }
        expect(
          searchItems({
            items: mockItems,
            searchablePaths: [conditionalFunction],
            searchValue: 'fancy',
          })
        ).toEqual([mockItems[0]])
      })

      it('does not return any items when conditions are not met', () => {
        const conditionalFunction = (item: any, searchValue: string) => {
          const category = _.get(item, 'category')
          const searchValueInLabel = _.includes(_.lowerCase(category), _.lowerCase(searchValue))

          if (searchValueInLabel && item.description === 'Incorrect!!!!') {
            return true
          }

          return false
        }

        expect(
          searchItems({
            items: mockItems,
            searchablePaths: [conditionalFunction],
            searchValue: 'fancy',
          })
        ).toHaveLength(0)
      })
    })
  })
})
