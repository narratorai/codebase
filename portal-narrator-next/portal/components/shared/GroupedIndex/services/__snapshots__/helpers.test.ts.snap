// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupedIndex/services/helpers #getAllPossibleGroupKeys gets all group keys including GROUP_RECENT keys 1`] = `
[
  "fancy",
  "hipster",
  "New",
  "Existing",
  "Uncategorized",
]
`;

exports[`GroupedIndex/services/helpers #getAllPossibleGroupKeys with groupLabelOverrides gets all group keys from overrides as well 1`] = `
[
  "Validated",
  "Not Validated",
  "New",
  "Existing",
  "Uncategorized",
]
`;

exports[`GroupedIndex/services/helpers #groupItems groups items 1`] = `
{
  "fancy": [
    {
      "category": "fancy",
      "description": "This is a description",
      "name": "Fancy Name",
      "nested": {
        "value": "nested is fun!",
      },
      "null_category": null,
      "only_in_fancy_name": "better only see me in fancy!",
      "other": "Should not be searched",
      "validated": false,
    },
  ],
  "hipster": [
    {
      "category": "hipster",
      "description": "Banjo anim ramps chia VHS batch flexitarian bespoke meh pinterest esse beard.",
      "name": "Hipster Ipsum",
      "nested": {
        "value": "Green juice cillum ut, 3 wolf moon",
      },
      "null_category": null,
      "other": "Should not be searched",
      "validated": false,
    },
  ],
}
`;

exports[`GroupedIndex/services/helpers #groupItems with groupLabelOverrides adds values from one group to both when passed addToGroup 1`] = `
{
  "fancy": [
    {
      "category": "fancy",
      "description": "This is a description",
      "name": "Fancy Name",
      "nested": {
        "value": "nested is fun!",
      },
      "null_category": null,
      "only_in_fancy_name": "better only see me in fancy!",
      "other": "Should not be searched",
      "validated": false,
    },
  ],
  "hipster": [
    {
      "category": "fancy",
      "description": "This is a description",
      "name": "Fancy Name",
      "nested": {
        "value": "nested is fun!",
      },
      "null_category": null,
      "only_in_fancy_name": "better only see me in fancy!",
      "other": "Should not be searched",
      "validated": false,
    },
    {
      "category": "hipster",
      "description": "Banjo anim ramps chia VHS batch flexitarian bespoke meh pinterest esse beard.",
      "name": "Hipster Ipsum",
      "nested": {
        "value": "Green juice cillum ut, 3 wolf moon",
      },
      "null_category": null,
      "other": "Should not be searched",
      "validated": false,
    },
  ],
}
`;

exports[`GroupedIndex/services/helpers #groupItems with groupLabelOverrides only overrides the default group names from _.groupBy 1`] = `
{
  "Not Validated": [
    {
      "category": "fancy",
      "description": "This is a description",
      "name": "Fancy Name",
      "nested": {
        "value": "nested is fun!",
      },
      "null_category": null,
      "only_in_fancy_name": "better only see me in fancy!",
      "other": "Should not be searched",
      "validated": false,
    },
    {
      "category": "hipster",
      "description": "Banjo anim ramps chia VHS batch flexitarian bespoke meh pinterest esse beard.",
      "name": "Hipster Ipsum",
      "nested": {
        "value": "Green juice cillum ut, 3 wolf moon",
      },
      "null_category": null,
      "other": "Should not be searched",
      "validated": false,
    },
  ],
}
`;

exports[`GroupedIndex/services/helpers #groupItems with null group values includes items with null groups into an "Uncategorized" group 1`] = `
{
  "Uncategorized": [
    {
      "category": "fancy",
      "description": "This is a description",
      "name": "Fancy Name",
      "nested": {
        "value": "nested is fun!",
      },
      "null_category": null,
      "only_in_fancy_name": "better only see me in fancy!",
      "other": "Should not be searched",
      "validated": false,
    },
    {
      "category": "hipster",
      "description": "Banjo anim ramps chia VHS batch flexitarian bespoke meh pinterest esse beard.",
      "name": "Hipster Ipsum",
      "nested": {
        "value": "Green juice cillum ut, 3 wolf moon",
      },
      "null_category": null,
      "other": "Should not be searched",
      "validated": false,
    },
  ],
}
`;
